### DISCUSSION TITLE
Tracking who a host connects to

### DISCUSSION THREAD

**Discussion by sporksmith** [shadow/shadow#2561] [Q&A]
@trinity-1686a is looking into extending our arti shadow CI to test the WIP bridge support, and would like to validate that a bridge-using client doesn't connect to tor nodes other than the bridge.

Our strace logging for the `connect` syscall currently just logs the pointer of the address. It'd be nice to fix that to parse and pretty-print the `struct sockaddr*`, but I think this might be a bit in flux with @stevenengler 's current work?

Another possibility is to record pcap files and use `tcpdump` with a packet filter. I gave that a quick try and `tcpdump` reports "invalid header" for every packet, presumably because of Shadow's TCP sequence numbering. I know we can work around this in wireshark by telling it to ignore sequence numbers, but don't immediately see such an option in `tcpdump`, and afaik `wireshark` isn't easily scriptable. Any ideas? It's been a bit since I've used `tcpdump` so maybe I'm missing something simple.

I tried adding pcap logging to tcp-blocking-lossless.yaml:

```
diff --git a/src/test/tcp/tcp-blocking-lossless.yaml b/src/test/tcp/tcp-blocking-lossless.yaml
index 34edf2732..3cb440dbb 100644
--- a/src/test/tcp/tcp-blocking-lossless.yaml
+++ b/src/test/tcp/tcp-blocking-lossless.yaml
@@ -18,6 +18,8 @@ network:
           packet_loss 0.0
         ]
       ]
+host_defaults:
+  pcap_directory: '.'
 hosts:
   lossless.tcpserver.echo:
     network_node_id: 0
```

And then get:

```
$ tcpdump -r build/src/test/tcp/tcp-blocking-lossless-shadow.data/hosts/lossless.tcpclient.echo/lossless.tcpclient.echo-11.0.0.1.pcap
reading from file build/src/test/tcp/tcp-blocking-lossless-shadow.data/hosts/lossless.tcpclient.echo/lossless.tcpclient.echo-11.0.0.1.pcap, link-type RAW (Raw IP), snapshot length 65535
[Invalid header: len(40) < caplen(52)]
[Invalid header: len(40) < caplen(52)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(40) < caplen(52)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(40) < caplen(52)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(1060) < caplen(1072)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(40) < caplen(52)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(40) < caplen(52)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(1500) < caplen(1512)]
[Invalid header: len(1060) < caplen(1072)]
[Invalid header: len(40) < caplen(52)]
[Invalid header: len(40) < caplen(52)]
[Invalid header: len(40) < caplen(52)]
[Invalid header: len(40) < caplen(52)]
[Invalid header: len(40) < caplen(52)]
```

**Comment by sporksmith** [comment]
We should be careful to look for UDP leaks as well; e.g. DNS. We could check the other relevant syscalls (`sendto`, `recvfrom`, ...), but maybe the pcap approach will ultimately be more robust.

DNS in particular is a bit tricky since shadow short-circuits them; a libc call to to "getaddrinfo" etc. won't typically be visible in the pcap file. Instead it makes a special syscall `shadow_hostname_to_addr_ipv4`; we should probably check the strace files to ensure the bridge client (or any tor client, really), doesn't unexpectedly make that syscall.

**Comment by sporksmith** [comment]
I've been directed to `tshark`, the terminal-based version of `wireshark`. It looks like it does indeed parse well enough to at least get the source and destination addresses:

```
$ tshark -r build/src/test/tcp/tcp-blocking-lossless-shadow.data/hosts/lossless.tcpclient.echo/lossless.tcpclient.echo-11.0.0.1.pcap
    1   0.000000     11.0.0.1 → 11.0.0.2     TCP 40 12739 → 1234 [SYN] Seq=1 Win=119[Malformed Packet]
    2   0.100000     11.0.0.2 → 11.0.0.1     TCP 40 1234 → 12739 [SYN, ACK] Seq=1 Ack=2 Win=119[Malformed Packet]
    3   0.100000     11.0.0.1 → 11.0.0.2     TCP 1500 12739 → 1234 [ACK] Seq=2 Ack=2 Win=897 Len=1448
    4   0.100000     11.0.0.1 → 11.0.0.2     TCP 1500 12739 → 1234 [ACK] Seq=3 Ack=2 Win=897 Len=1448
    5   0.201000     11.0.0.2 → 11.0.0.1     TCP 40 1234 → 12739 [ACK] Seq=0 Ack=4 Win=119[Malformed Packet]
    6   0.201000     11.0.0.1 → 11.0.0.2     TCP 1500 12739 → 1234 [ACK] Seq=4 Ack=2 Win=897 Len=1448
    7   0.201000     11.0.0.1 → 11.0.0.2     TCP 1500 12739 → 1234 [ACK] Seq=5 Ack=2 Win=897 Len=1448
    8   0.201000     11.0.0.1 → 11.0.0.2     TCP 1500 12739 → 1234 [ACK] Seq=6 Ack=2 Win=897 Len=1448
    9   0.201000     11.0.0.1 → 11.0.0.2     TCP 1500 12739 → 1234 [ACK] Seq=7 Ack=2 Win=897 Len=1448
   10   0.301000     11.0.0.2 → 11.0.0.1     TCP 40 1234 → 12739 [ACK] Seq=0 Ack=8 Win=897[Malformed Packet]
   11   0.301000     11.0.0.1 → 11.0.0.2     TCP 1500 12739 → 1234 [ACK] Seq=8 Ack=2 Win=897 Len=1448
   12   0.301000     11.0.0.1 → 11.0.0.2     TCP 1500 12739 → 1234 [ACK] Seq=9 Ack=2 Win=897 Len=1448
   13   0.301000     11.0.0.1 → 11.0.0.2     TCP 1500 12739 → 1234 [ACK] Seq=10 Ack=2 Win=897 Len=1448
   14   0.301000     11.0.0.1 → 11.0.0.2     TCP 1500 12739 → 1234 [ACK] Seq=11 Ack=2 Win=897 Len=1448
   15   0.301000     11.0.0.1 → 11.0.0.2     TCP 1500 12739 → 1234 [ACK] Seq=12 Ack=2 Win=897 Len=1448
   16   0.301000     11.0.0.1 → 11.0.0.2     TCP 1500 12739 → 1234 [ACK] Seq=13 Ack=2 Win=897 Len=1448
   17   0.301000     11.0.0.1 → 11.0.0.2     TCP 1500 12739 → 1234 [ACK] Seq=14 Ack=2 Win=897 Len=1448
   18   0.301000     11.0.0.1 → 11.0.0.2     TCP 1060 12739 → 1234 [ACK] Seq=15 Ack=2 Win=897 Len=1008
   19   0.401000     11.0.0.2 → 11.0.0.1     TCP 1500 1234 → 12739 [ACK] Seq=2 Ack=16 Win=897 Len=1448
   20   0.401000     11.0.0.2 → 11.0.0.1     TCP 1500 1234 → 12739 [ACK] Seq=3 Ack=16 Win=897 Len=1448
   21   0.402000     11.0.0.1 → 11.0.0.2     TCP 40 12739 → 1234 [ACK] Seq=0 Ack=4 Win=897[Malformed Packet]
   22   0.502000     11.0.0.2 → 11.0.0.1     TCP 1500 1234 → 12739 [ACK] Seq=4 Ack=16 Win=897 Len=1448
   23   0.502000     11.0.0.2 → 11.0.0.1     TCP 1500 1234 → 12739 [ACK] Seq=5 Ack=16 Win=897 Len=1448
   24   0.502000     11.0.0.2 → 11.0.0.1     TCP 1500 1234 → 12739 [ACK] Seq=6 Ack=16 Win=897 Len=1448
   25   0.502000     11.0.0.2 → 11.0.0.1     TCP 1500 1234 → 12739 [ACK] Seq=7 Ack=16 Win=897 Len=1448
   26   0.503000     11.0.0.1 → 11.0.0.2     TCP 40 12739 → 1234 [ACK] Seq=0 Ack=8 Win=897[Malformed Packet]
   27   0.603000     11.0.0.2 → 11.0.0.1     TCP 1500 1234 → 12739 [ACK] Seq=8 Ack=16 Win=897 Len=1448
   28   0.603000     11.0.0.2 → 11.0.0.1     TCP 1500 1234 → 12739 [ACK] Seq=9 Ack=16 Win=897 Len=1448
   29   0.603000     11.0.0.2 → 11.0.0.1     TCP 1500 1234 → 12739 [ACK] Seq=10 Ack=16 Win=897 Len=1448
   30   0.603000     11.0.0.2 → 11.0.0.1     TCP 1500 1234 → 12739 [ACK] Seq=11 Ack=16 Win=897 Len=1448
   31   0.603000     11.0.0.2 → 11.0.0.1     TCP 1500 1234 → 12739 [ACK] Seq=12 Ack=16 Win=897 Len=1448
   32   0.603000     11.0.0.2 → 11.0.0.1     TCP 1500 1234 → 12739 [ACK] Seq=13 Ack=16 Win=897 Len=1448
   33   0.603000     11.0.0.2 → 11.0.0.1     TCP 1500 1234 → 12739 [ACK] Seq=14 Ack=16 Win=897 Len=1448
   34   0.603000     11.0.0.2 → 11.0.0.1     TCP 1060 1234 → 12739 [ACK] Seq=15 Ack=16 Win=897 Len=1008
   35   0.603000     11.0.0.1 → 11.0.0.2     TCP 40 12739 → 1234 [FIN] Seq=16 Win=897[Malformed Packet]
   36   0.604000     11.0.0.1 → 11.0.0.2     TCP 40 12739 → 1234 [ACK] Seq=0 Ack=16 Win=897[Malformed Packet]
   37   0.703000     11.0.0.2 → 11.0.0.1     TCP 40 1234 → 12739 [FIN, ACK] Seq=0 Ack=16 Win=897[Malformed Packet]
   38   0.704000     11.0.0.2 → 11.0.0.1     TCP 40 1234 → 12739 [FIN] Seq=16 Win=897[Malformed Packet]
   39   0.704000     11.0.0.1 → 11.0.0.2     TCP 40 12739 → 1234 [FIN, ACK] Seq=0 Ack=16 Win=897[Malformed Packet]
```

**Comment by stevenengler** [comment]
> Our strace logging for the `connect` syscall currently just logs the pointer of the address. It'd be nice to fix that to parse and pretty-print the `struct sockaddr*`, but I think this might be a bit in flux with @stevenengler 's current work?

The problem here is that you can't print a `struct sockaddr*` safely by itself. You also need the address length field as well, and the syscall logging code isn't currently designed to display types that need multiple syscall arguments to parse them. The last time it was brought up we decided we didn't really need it at the time, but if this is something that would be useful, I think I have some ideas on how to implement this.

**Reply by sporksmith** [reply]
Ahh right.

After thinking through this some more, I think validating the pcap files will be a more robust long-term solution anyway. e.g. it'd be a shame if we put this test in place to look for `connect` and `sendto`, and then next year after we've forgotten the CI was doing this, Linux introduces a `connect2` syscall, and shadow adds support for it, but our CI test now has a gap...

So yeah, maybe we can kick the can a while longer on pretty printing in the strace logger.

**Reply by sporksmith** [reply]
If we did want to fix the logging though, and if we're doing the logging from the shadow process, I think we *could* use e.g. `memory_ref_prefix` to read up to `sizeof(sockaddr_storage)` or whatever. Even if the managed process hasn't initialized the suffix of that region (or padding), from the perspective of the kernel, and what shadow gets back, they should be bytes with a fixed (but arbitrary) value, not Undefined.

OTOH if you have a plan to eventually add support for the logging to be able to see the supplied length parameter, that'd probably be preferable.

**Reply by stevenengler** [reply]
This would work for `sockaddr_in`, but not for `sockaddr_un` since the length determines the address. For example, `let addr: libc::sockaddr_un = std::mem::zeroed(); addr.sun_family = AF_UNIX;` with an address length of 10 would be a different address than the same struct with an address length of 11, so there wouldn't be a nice way to display them. I'm not against adding specific cases in, for example only displaying/printing a `sockaddr_in` and not a `sockaddr_un`, but just wanted to add an example this wouldn't work for.

**Comment by trinity-1686a** [comment]
FYI I ended up generating pcaps and processing them with `tshark`, and looking for unexpected `shadow_hostname_to_addr_ipv4` in strace logs. It seems to be working fine
